src/Lean/Server/InfoUtils.lean

```diff
@@ -160,18 +159,21 @@ structure GoalsAtResult where
   - None of the `children` can provide satisfy the condition above. That is, for composite tactics such as
     `induction`, we always give preference for information stored in nested (children) tactics.
 
-  Moreover, we instruct the LSP server to use the state after the tactic execution if hoverPos >= endPos
+  Moreover, we instruct the LSP server to use the state after the tactic execution if hoverPos >= endPos *and*
+  there is no nested tactic info (i.e. it is a leaf tactic; tactic combinators should decide for themselves
+  where to show intermediate/final states)
```

Comment:
The logic for this (see below) is not quite complete yet; it doesn't work for `match` since there is a macro step in between, meaning you will see the final state (usually "no goals") if you put the cursor anywhere between `match` and `=>`. But it is definitely an improvement over the current state for many other tactics, in particular `induction`.

Reply:
fixed